### PR TITLE
Updated TF attack

### PR DESCRIPTION
**Pull request by bzamanlooy**
# PR Type
[Feature | Fix | Documentation | Other ]
Fix
# Short Description

This pull request is doing three things:

1. Modifying the Clavaddpm setup to allow for profited label encoders to be passed during training.
2. Modifying the TF attack to figure out the Gaussian noise dimension from the data. 
3. Modifying the TF attack to not refit transformations when loading tables.

# Tests
- Updated there TF attack tests to reflect a more realistic setup

### CODE REVIEW

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py [hunk 1]
```diff
@@ -95,14 +100,47 @@ def encode_and_merge_features(
...
+                f"{sorted(missing_cols)}. "
+                "Refusing to mix preloaded encoders with freshly fit encoders."
+            )
+
     categorical_data_encoded = []
     label_encoders = {}
     for column in range(all_categorical_data.shape[1]):
-        label_encoder = LabelEncoder()
-        encoded_labels = label_encoder.fit_transform(all_categorical_data[:, column]).astype(float)
+        if preloaded_encoders is not None:
```
Comment (bzamanlooy) [inline src/midst_toolkit/models/clavaddpm/dataset_utils.py:137]: @emersodb quick question. In 113 and 114 I have asserted that categorical_column_names is not None but because of the mypy error, I also added it to 134. Should I have just have ignored the error?

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 import csv
 import os
 from collections.abc import Generator
+from dataclasses import replace as dc_replace
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: Any reason for this import rename rather than just using the `replace` name directly?

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 3]
```diff
@@ -98,6 +99,7 @@ def mixed_loss(
 
 
 # TODO: Unify this with the Dataset.from_df function.
+# TODO: Noise scale is always called with a value of 0 for the attack.
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: For this TODO, I think this is describing what is happening in the call but not necessarily what we want to do in the todo (😂)

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 4]
```diff
@@ -108,27 +110,26 @@ def make_dataset_from_df_with_loaded(
...
-        transformation: Transformations that one might apply to the dataset, including NaN policies etc.
-        is_target_conditioned: Enum indicating how, if at all, the model uses a target for generation conditioning.
-        table_metadata: Meta data about the table or tables.
-        label_encoders: Encoders that were used to encode the categorical data.
-        numerical_transform: Transformations that should be applied to the numerical data. Defaults to None.
-        noise_scale: he scale of the noise to add to the categorical features. Noise is drawn from a normal
-            distribution with standard deviation of ``noise_scale``. Defaults to 0.
+        data: The dataframe to make the dataset from.
+        transformation: The transformations to apply to the data.
+        is_target_conditioned: Whether the target is conditioned on the data.
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: I think I like the former description of this argument better? Is there a reason you when with the one you have here?

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 5]
```diff
@@ -108,27 +110,26 @@ def make_dataset_from_df_with_loaded(
...
+        table_metadata: The metadata for the table.
+        label_encoders: The label encoders for the categorical columns.
+        numerical_transform: The numerical transform to apply to the data.
+        noise_scale: The scale of the noise to add to the data.
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: Any reason we went with the simplified documentation here?

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 6]
```diff
@@ -153,6 +154,13 @@ def make_dataset_from_df_with_loaded(
         numerical_features = categorical_features
 
     target_info = TargetInfo(policy=None, mean=None, std=None)
+
+    # Apply the model's pre-fitted numerical transform directly instead of re-fitting a new one.
+    # Calling transform_dataset() would fit a brand new QuantileTransformer on the MIA data,
+    # which produces a different normalization than the model saw during training, destroying signal.
+    if numerical_transform is not None:
+        numerical_features = {k: numerical_transform.transform(v) for k, v in numerical_features.items()}
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py:164]: This is slightly hacky, but with the current code structure, I don't think you have a choice. The transform dataset should really determine whether the the scaler has been fitted and if it has, then apply it rather than just fitting a new one.

Comment (emersodb) [reply src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py:164]: I would say to edit the transform dataset code to do this, but that might be opening a horrible can of worms.

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 7]
```diff
@@ -448,7 +458,28 @@ def train_tartan_federer_attack_classifier(
     population_df_for_validation = pd.read_csv(population_data_dir / "population_dataset_for_validating_attack.csv")
     log(INFO, "Population datasets for validating loaded.")
 
-    noise_dimension = len([col for col in population_df_for_training.columns if "_id" not in col])
+    # Fix 1: derive noise dimension from the actual diffusion model's num_numerical_features rather
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: I would drop the Fix 1: prefix here. The documentation is useful to describe what you're doing.

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 8]
```diff
@@ -448,7 +458,28 @@ def train_tartan_federer_attack_classifier(
...
+    # than from the population dataframe column count.  The mixed_loss function slices
+    # x[:, :diffusion.num_numerical_features], so the noise vectors must have exactly that length.
+    # We load the first available model to read this value, then discard it.
+    first_model_number = (train_indices + (val_indices or []))[0]
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: This is perhaps a naive question, but train_indices can't really be empty right? Why not just grab `train_indices[0]` instead?

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 9]
```diff
@@ -448,7 +458,28 @@ def train_tartan_federer_attack_classifier(
...
+    first_model_dir = model_data_dir / f"{model_type}_{first_model_number}"
+    first_model_path = first_model_dir / target_model_subdir
+
+    if model_type != "tabddpm":
+        raise ValueError(
+            f"Unsupported model_type {model_type}. Tartan Federer Attack is only supported for ClavaDDPM-single-table models."
+        )
+    # TODO: We should read this from the metadata instead.
+    _relation_order = [("None", "trans")]
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: I may be missing something, but since this is a hardcoded tuple, why are we bothering with the for loop? That is, can we simply forgo it and define _parent = "None" and _child = "trans" and do the stuff inside as a oneoff?

File: src/midst_toolkit/models/clavaddpm/dataset.py [hunk 10]
```diff
@@ -377,10 +380,14 @@ def from_df(
         column_orders = numerical_column_names + categorical_column_names
 
         # Encode the categorical features and merge them with the numerical features
+        # Look for pre-fitted label encoders in the parent directories of the data
+
         features, label_encoders = encode_and_merge_features(
             categorical_features,
             numerical_features,
             noise_scale,
+            categorical_column_names=categorical_column_names if len(categorical_column_names) > 0 else None,
```
Comment (emersodb) [inline src/midst_toolkit/models/clavaddpm/dataset.py]: I'd say don't do this here. Just send the `categorical_column_names` as is and handle empty lists in the function itself.

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py [hunk 11]
```diff
@@ -95,14 +100,47 @@ def encode_and_merge_features(
         )
     )
 
+    # Load pre-fitted label encoders from pkl if provided, otherwise fit on current data
+    preloaded_encoders: dict[str, LabelEncoder] | None = None
+    if label_encoders_path is not None:
+        _pkl_path = Path(label_encoders_path)
```
Comment (emersodb) [inline src/midst_toolkit/models/clavaddpm/dataset_utils.py:112]: leading underscores are typically not necessary. I'd just use `pkl_path` here unless I'm missing something?

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py [hunk 12]
```diff
@@ -95,14 +100,47 @@ def encode_and_merge_features(
...
+
+    if preloaded_encoders is not None:
+        assert categorical_column_names is not None, (
+            "categorical_column_names must be provided when using label_encoders_path."
+        )
+
+        expected_cols = set(categorical_column_names)
+        available_cols = set(preloaded_encoders.keys())
+
+        missing_cols = expected_cols - available_cols
```
Comment (emersodb) [inline src/midst_toolkit/models/clavaddpm/dataset_utils.py:127]: If the `categorical_column_names` is empty and that is unexpected, you'll throw here anyway. So I don't think you need the 
```
categorical_column_names if len(categorical_column_names) > 0 else None
```
from above anyway?

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py [hunk 13]
```diff
@@ -95,14 +100,47 @@ def encode_and_merge_features(
...
+
+        if missing_cols:
+            raise ValueError(
+                "label_encoders_path is missing encoders for categorical columns: "
+                f"{sorted(missing_cols)}. "
+                "Refusing to mix preloaded encoders with freshly fit encoders."
```
Comment (emersodb) [inline src/midst_toolkit/models/clavaddpm/dataset_utils.py]: I'd say be less specific here. What you're doing is refusing a mis-match between the provided categorical names and the label encoder keys.

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py (hunk 1 above)
Comment (emersodb) [reply src/midst_toolkit/models/clavaddpm/dataset_utils.py:137]: That assertion only gets hit if `preloaded_encoders` is not None above. Mypy can't trace the implications of the combinations of the if statement and the assertion (that's an execution path thing). So you have to do it again here.

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py [hunk 14]
```diff
@@ -95,14 +100,47 @@ def encode_and_merge_features(
         )
     )
 
+    # Load pre-fitted label encoders from pkl if provided, otherwise fit on current data
+    preloaded_encoders: dict[str, LabelEncoder] | None = None
```
Comment (emersodb) [inline src/midst_toolkit/models/clavaddpm/dataset_utils.py]: I will note that I was very confused here. Your label encoder dictionary here is index by strings (column names) by the label_encoders that are to be returned are index by column indices. You do end up taking care of this later on. However, I would suggest you add a comment here to explain what's happening because the label encoder you're preloading is definitely not of the same "kind" as the ones we are constructing here (that is, it must be constructed somewhere else, we're not reusing an artifact formed by the process)

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py [hunk 15]
```diff
@@ -75,6 +77,9 @@ def encode_and_merge_features(
             keys are "train", "val", "test" from the DataSplit enumeration
         noise_scale: The scale of the noise to add to the categorical features. Noise is drawn from a normal
             distribution with standard deviation of ``noise_scale``.
+        categorical_column_names: The names of the categorical columns.
+        label_encoders_path: The path to the label encoders pkl file. If provided, already fitted label encoder
```
Comment (emersodb) [inline src/midst_toolkit/models/clavaddpm/dataset_utils.py]: I would add a comment as to why we might do this.

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 16]
```diff
@@ -448,7 +458,28 @@ def train_tartan_federer_attack_classifier(
...
+    for _parent, _child in _relation_order:
+        _ckpt_path = first_model_path / f"{_parent}_{_child}_ckpt.pkl"
+        with open(_ckpt_path, "rb") as _f:
+            _probe_model = CustomUnpickler(_f).load()
+        noise_dimension = _probe_model.diffusion.num_numerical_features
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: I'm struggling to find where this property is being pulled from. Do you mind directing me to the right spot in the codebase?

**Review by emersodb — COMMENTED** [review]
Thanks for putting these fixes into a PR! Most of my comments are pretty small. Some of the hoops you're jumping through is a result of inflexible code elsewhere, which we should fix at some other point.

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py (hunk 3 above)
Comment (bzamanlooy) [reply src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: I added an explanation.

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py (hunk 8 above)
Comment (bzamanlooy) [reply src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: Yeah makes a lot of sense.

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py (hunk 9 above)
Comment (bzamanlooy) [reply src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: Yep makes sense. I kept the _relation_order to match the way we keep it in the metadata but simplify the code so no loop is used.

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py (hunk 13 above)
Comment (bzamanlooy) [reply src/midst_toolkit/models/clavaddpm/dataset_utils.py]: makes sense.

File: src/midst_toolkit/models/clavaddpm/dataset_utils.py (hunk 14 above)
Comment (bzamanlooy) [reply src/midst_toolkit/models/clavaddpm/dataset_utils.py]: Yeah fair enough I will do that.

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 17]
```diff
@@ -99,7 +99,8 @@ def mixed_loss(
 
 
 # TODO: Unify this with the Dataset.from_df function.
-# TODO: Noise scale is always called with a value of 0 for the attack.
+# TODO: Noise scale is always called with a value of 0 for the attack. So we should remove it from the f
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: I think the f here is a typo?

File: src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py [hunk 18]
```diff
@@ -472,13 +474,12 @@ def train_tartan_federer_attack_classifier(  # noqa: PLR0915, PLR0912
...
-        _ckpt_path = first_model_path / f"{_parent}_{_child}_ckpt.pkl"
-        with open(_ckpt_path, "rb") as _f:
-            _probe_model = CustomUnpickler(_f).load()
-        noise_dimension = _probe_model.diffusion.num_numerical_features
-        log(INFO, f"Noise dimension read from diffusion model: {noise_dimension}")
-        break
+    _parent, _child = _relation_order[0]
+    _ckpt_path = first_model_path / f"{_parent}_{_child}_ckpt.pkl"
+    with open(_ckpt_path, "rb") as _f:
+        _probe_model = CustomUnpickler(_f).load()
```
Comment (emersodb) [inline src/midst_toolkit/attacks/tartan_federer/tartan_federer_attack.py]: Any reason we're using the `_` prefixes here? I'd say drop them unless they are serving a purpose that I'm missing 🙂

**Review by emersodb — APPROVED** [review]
Changes look good. Just two very small comments.
